activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -465,6 +467,7 @@ def replace_on_target(record, skip_callbacks, replace:, inversing: false)
             target[index] = record
```

Comment:
Shouldn't we set `@target_index_map[record]` here too?

There's also a few places in this file where `@target` is set directly, but it's not clear to me if `@target_index_map` actually needs to be reset in those cases.